### PR TITLE
Rpc

### PR DESCRIPTION
**Pull request by tiram88**
This is basically a copy of the code developed in [rpc-poly-prototype](https://github.com/tiram88/poly-rpc-prototype) with some minor adjustments, notably `Uint192` because of its use as `BlueWorkType`.

This PR features:
- Trait `RpcApi` (initial work of [Aspectron](https://github.com/aspectron/rusty-kaspa/blob/ffa8dd95264e68580acabb6cccb6a37dfd39c30a/rpc/core/src/client/interface.rs)) exposing a protocol-agnostic API
- rpc-core `RpcApi` server responsible for fetching data from consensus
- rpc-grpc `RpcApi` service and client implementing (partially) the `protowire` RPC from kaspad
- Notification system with following pipeline: consensus -> rpc-core server -> rpc-grpc service -> rpc-grpc client -> client code

### Static architecture of `Notifier`

The design has 4 main structs providing the required functionality:

1) `Notifier`: registers/unregisters listeners, start/stop notification types for them and notify them of incoming notifications
2) `Listener`: receives notifications
3) `Collector`: receives notifications of any type from a channel, converts them into `rpc_core::Notification` and send them to a `Notifier`
4) `Subscriber`: provides a retro-action loop, configuring the source of a `Notifier` (any struct implementing the trait `SubscriptionManager`) with the actual needs of the `Notifier` based on which notification type its listeners actually started / stopped

The `Notifier` has a map of `Listener`s and, optionally, one `Collector` and one `Subscriber`.

This design offers a maximum of flexibility for all use cases (consensus to rpc-core, rpc-core to grpc server, grpc-client to rpc-core, any other future protocol implementation). It also offers a non-blocking async flow, with channels when useful and strong type coupling where appropriate.

Every of the 3 `RpcApi` implementations has its own `Notifier`.

### Pipeline design & Data flow

The notification pipeline consists of following steps (from top to bottom):

- consensus
- rpc-core server
- rpc-grpc server
- rpc-grpc client
- client-side external code.

Basically we have notifications travelling/amplifying downstream and listener start_/stop_notify making subscription requests travel/concentrate upstream.

#### consensus

For now, no notifier here (because no actual notification source). We only have a `Notification` enum based on consensus structs and some consensus to rpc-core notification converters used by the next step.

#### rpc-core

rpc-core server `RpcApi` implementation has a `Notifier` and a `Collector` able to convert `ConsensusNotification` into rpc-core `Notification`. It has no `Subscriber` yet (some consensus notification source has to be implemented).

#### rpc-grpc server

rpc-grpc server has a service implementing `RpcApi`. The service registers as a listener of rpc-core service and calls it as a subscription manager to let the latter know what it’s own listeners need.

Into more details, the service uses rpc-core server, has a `Notifier`, a `Collector` and a `Subscriber`. The `Collector` has a `Listener` of rpc-core server `Notifier`. The `Subscriber` uses rpc-core server `Notifier` as subscription manager. Additionally, the service has a `GrpcConnectionManager` responsible for managing the gRPC connections with clients.

Every incoming gRPC connection to the service creates a `GrpcConnection` which becomes a listener of the service `Notifier`. Every notify message (ie. `NotifyBlockAddedRequest`) received by the gRPC service turns into a notifier start_/stop_notify call which sends an Add-/RemoveListener message to a matching notifier dispatch worker, which in turn may, if needed, trigger a matching suscription (via Start/Stop `SubscribeCommand`) to rpc-core service.

#### rpc-grpc client

The rpc-grpc client has a `Notifier` and a `Resolver` (might deserve a better name).

Into more details, the client `Notifier` has a `Collector` getting notifications from the `Resolver` and a `Subscriber` using the `Resolver` as `SubscriptionManager`.

The `Resolver`'s main job is to send request messages to a `protowire` server and pair those with incoming response messages from the same server. This occurs with the help of `Pending` instances. Another responsibility of `Resolver` is to handle the notification flow, transforming and forwarding the incoming notification messages to the client `Notifier` and acting for it as a `SubscriptionManager` relaying the subscription start_/stop_notify calls to the gRPC server sending it `Notify...Request` messages.

#### client-side external code

Here the code referencing the rpc-grpc client crate can instantiate a client, register a set of listeners, subscribe them to some notification types so that they get the matching notifications.

Seen by the external code, the whole RPC sub-system is essentially only exposing an instance of a struct implementing the `RpcApi` trait (plus some protocol-specific constructor/connector). This makes switching to other (to come) protocols seamless and was the ultimate requirement for this whole architecture.

## Limitations

The gRPC implementation at this stage should be considered essentially a proof of concept. It is limited to 2 queries: `get_block` and `get_info`.

Unit-tests must be added particularly (but not only) for converters. Some integration test should be envisioned too.

The same is true of the notification system that only implements `BlockAdded`.

The whole rpc sub-system is not actually connected to consensus for now, neither for queries nor for notifications.

The rpc sub-system is by no means production-ready. Some processes still need a decoupling with a queue of messages and a pool of workers, bounded channels should be introduced in some places and backpressure should correctly addressed.

All these issues will have further dedicated PRs.

### CODE REVIEW

**Comment by KaffinPX** [comment]
plz no gRPC

**Comment by tiram88** [comment]
gRPC is the minimal RPC requirement for inter-operability and backwards compatibility. However, we have plans for 2 more protocols: borsh and json. So most probable use-cases should all be covered.

File: rpc/core/src/notify/subscriber.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        if !self.subscribe_is_running.load(Ordering::SeqCst) {
+            let (shutdown_trigger, shutdown_listener) = triggered::trigger();
+            let mut subscribe_shutdown_listener = self.subscribe_shutdown_listener.lock().unwrap();
+            *subscribe_shutdown_listener = Some(shutdown_listener);
+            self.subscribe_task(shutdown_trigger, self.subscribe_channel.receiver());
+        }
+    }
+
+    /// Launch the subscribe task
+    fn subscribe_task(&self, shutdown_trigger: triggered::Trigger, subscribe_rx: Receiver<SubscribeMessage>) {
```
Comment (someone235) [inline rpc/core/src/notify/subscriber.rs:72]: subscribe_loop?

File: rpc/core/src/notify/subscriber.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        let subscribe_is_running = self.subscribe_is_running.clone();
+        subscribe_is_running.store(true, Ordering::SeqCst);
+        let subscription_manager = self.subscription_manager.clone();
+        // let listener = self.listener.clone();
```
Comment (someone235) [inline rpc/core/src/notify/subscriber.rs:76]: Can you remove comment?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -53,3 +57,6 @@ wasm-bindgen = { version = "0.2", features = ["serde-serialize"] }
 criterion = { version = "0.4", default-features = false }
 indexmap = "1.9.1"
 smallvec = { version = "1.10.0", features = ["serde"] }
+borsh = "0.9.3"
```
[the remark this answers is not in the document]
Comment (tiram88) [reply Cargo.toml:60]: It is a perhaps good thing if we can do that.

But then what about Anton's ongoing development of the RPC borsh client/server?

[the remark this answers is not in the document]
Comment (tiram88) [reply Cargo.toml]: So as discussed yesterday, I remove borsh as a workspace dependency and keep it in rpc-core

**Review by tiram88 — COMMENTED** [review]
Thank you @someone235 for your review.

I addressed all issues. Quite a lot of changes, as expected.

**Review by someone235 — CHANGES_REQUESTED** [review]

File: rpc/grpc/src/server/connection.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+use tokio::sync::mpsc;
+
+pub type GrpcSender = mpsc::Sender<StatusResult<KaspadResponse>>;
+
+pub(crate) struct GrpcConnection {
+    _address: SocketAddr,
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/grpc/src/server/connection.rs]: Fixed

File: rpc/grpc/src/server/service.rs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            self.core_service.stop_notify(listener_id, event.into()).await?;
+        }
+
+        // Stop the internal notifier
+        self.notifier.clone().stop().await?;
+
+        Ok(())
+    }
+
+    pub async fn finalize(&self) -> RpcResult<()> {
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/grpc/src/server/service.rs:104]: yes

**Review by tiram88 — COMMENTED** [review]
Everything addressed

**Review by someone235 — APPROVED** [review]

**Pull request merged**